tt_eager/tt_dnn/op_library/bmm/multi_core_reuse_mcast_2d_optimized/bmm_op_multi_core_reuse_mcast_2d_optimized.cpp

```diff
@@ -113,77 +112,145 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
-        {(std::size_t)start_core_x + num_cores_c - 1, (std::size_t)start_core_y + num_cores_r - 1});
+        {(std::size_t)start_core_x + num_cores_with_work_c - 1, (std::size_t)start_core_y + num_cores_with_work_r - 1});
 
     CoreRange all_except_top_row(
         {(std::size_t)start_core_x, (std::size_t)start_core_y + 1},
-        {(std::size_t)start_core_x + num_cores_c - 1, (std::size_t)start_core_y + num_cores_r - 1});
+        {(std::size_t)start_core_x + num_cores_with_work_c - 1, (std::size_t)start_core_y + num_cores_with_work_r - 1});
 
     CoreRange left_column_except_corner(
         {(std::size_t)start_core_x, (std::size_t)start_core_y + 1},
-        {(std::size_t)start_core_x, (std::size_t)start_core_y + num_cores_r - 1});
+        {(std::size_t)start_core_x, (std::size_t)start_core_y + num_cores_with_work_r - 1});
 
     CoreRange top_row_except_corner(
         {(std::size_t)start_core_x + 1, (std::size_t)start_core_y},
-        {(std::size_t)start_core_x + num_cores_c - 1, (std::size_t)start_core_y});
+        {(std::size_t)start_core_x + num_cores_with_work_c - 1, (std::size_t)start_core_y});
 
+    ////////////////////////////////////////////////////////////////////////////
+    //                      INTERLEAVED SENDER/RECEIVER
+    ////////////////////////////////////////////////////////////////////////////
     CoreRange in0_sender = left_column;
     CoreRange in0_sender_in1_receiver = left_column_except_corner;
     CoreRange in1_sender = top_row;
     CoreRange in0_receiver_in1_sender = top_row_except_corner;
-
-    uint32_t in0_end = num_cores_r - 1;
-    uint32_t in1_end = num_cores_c - 1;
-
-    // in1 is the reader of weights/output writer, and we choose to make it use the optimized reader noc
-    tt_metal::NOC in0_noc = detail::GetPreferredNOCForDRAMWrite(device->arch());
-    tt_metal::NOC in1_noc = detail::GetPreferredNOCForDRAMRead(device->arch());
-    tt_metal::NOC in0_split_noc = detail::GetPreferredNOCForDRAMRead(device->arch());
-    tt_metal::NOC in1_split_noc = detail::GetPreferredNOCForDRAMWrite(device->arch());
     if (transpose_mcast) {
         std::swap(in0_sender, in1_sender);
         std::swap(in0_sender_in1_receiver, in0_receiver_in1_sender);
-        std::swap(in0_end, in1_end);
     }
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                      IN0 SHARDED SENDER/RECEIVER
+    ////////////////////////////////////////////////////////////////////////////
+    uint32_t num_cores_c = num_cores_with_work_c;
+    uint32_t num_cores_r = num_cores_with_work_r;
+    uint32_t in0_mcast_receiver_grid_diff_coord_start;
+    uint32_t in0_mcast_receiver_grid_diff_coord_end;
+    std::vector<uint32_t> in0_mcast_noc_x;
+    std::vector<uint32_t> in0_mcast_noc_y;
+    uint32_t in0_sender_num_cores_along_width;
+
+    // Only used for in0 block sharded
+    CoreRange in0_mcast_cores_with_work_and_in_receiver_grid(
```

Comment:
Instead of these two placeholders, why not define these variables, with the right values set from the start, further down in the if block you added around line 462 `if (in0_block_sharded) {` which at first glance looks like the only code that uses them?

Reply:
We can definitely move code like this together inside the same namespace. Problem is, in our c++ op host code, we typically have chunks of code that does something. In this case, we have:
- set up some `CoreRangeSet`s to describe regions on our device
- set up bunch of compile time args for reader kernels
- create the kernels using `CoreRangeSet` and compile time args

I think this makes the code a bit more readable. OR, we can move everything together (so fully create each kernel for each case). That also makes sense to me, but would probably be better in a separate PR.

Reply:
Ok. Leaving it as is in this PR is fine.